Add allowlistedCertificates field to TrustConfig

Hamzawy63 opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
This CL adds allowlistedCertificates field to TrustConfig resource.
Fixes https://github.com/hashicorp/terraform-provider-google/issues/18170 and https://github.com/hashicorp/terraform-provider-google/issues/18385

Tests that I run: 
`make testacc TEST=./google/services/certificatemanager TESTARGS='-run=TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample'`

`make testacc TEST=./google/services/certificatemanager TESTARGS='-run=TestAccCertificateManagerTrustConfig_update'`

Manual tests that I performed: 
- Create a resource with one allowlisted
- Add one more cert to the resource
- Remove one cert from the resoruce

<!--

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
certificatemanager: added `allowlisted_certificates` to `google_certificate_manager_trust_config`
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 154 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 154 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10906-old..auto-pr-10906) ( 1 file changed, 32 insertions(+))
Open in Cloud Shell: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 113 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: `20`
Passed tests: `17`
Skipped tests: `2`
Affected tests: `1`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>certificatemanager</li></ul></blockquote></details>


#### Action taken
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample </blockquote></details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/46302454-b4af-470a-89be-5e705688836e/recording/TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample.log)]


$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$

---
$\textcolor{green}{\textsf{All tests passed!}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/46302454-b4af-470a-89be-5e705688836e/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/46302454-b4af-470a-89be-5e705688836e/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 154 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 154 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10906-old..auto-pr-10906) ( 1 file changed, 32 insertions(+))
Open in Cloud Shell: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 113 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: `20`
Passed tests: `18`
Skipped tests: `2`
Affected tests: `0`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>certificatemanager</li></ul></blockquote></details>

$\textcolor{green}{\textsf{All tests passed!}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/edc36e0f-ffed-41f8-9ab7-7bc54278bdfe/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 162 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 162 insertions(+))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10906-old..auto-pr-10906) ( 1 file changed, 32 insertions(+))
Open in Cloud Shell: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 117 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: `20`
Passed tests: `17`
Skipped tests: `2`
Affected tests: `1`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>certificatemanager</li></ul></blockquote></details>


#### Action taken
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample </blockquote></details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/653305c9-5a02-447e-8474-ffb1d8d879f1/recording/TestAccCertificateManagerTrustConfig_certificateManagerTrustConfigAllowlistedCertificatesExample.log)]


$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$

---
$\textcolor{green}{\textsf{All tests passed!}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/653305c9-5a02-447e-8474-ffb1d8d879f1/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/653305c9-5a02-447e-8474-ffb1d8d879f1/recording) for each test

Hamzawy63 commented:
During my manual testing, I faced a strange behavior: If I make two consecutive update to the resource, the first update will be in-place (as expected) and the second one will force replace the resource. 
That was my resource definition: 
```
resource "google_certificate_manager_trust_config" "default" {
  name        = "sample-trust-configs"
  description = "A sample trust config resource with allowlisted certificates"
  location = "global"

  allowlisted_certificates  {
    pem_certificate = file("cert.pem") 
  }
  labels = {
    foo = "bar"
  }
}
````
The first update, I changed cert.pem to cert2.pem and the update was in-place and succssful 
The second update, I changed cert2.pem back to cert.pem. In this case, terraform decided to force replace the resource because it detected a change in effective_labels variable

```
  # google_certificate_manager_trust_config.default must be replaced
-/+ resource "google_certificate_manager_trust_config" "default" {
      ~ create_time      = "2024-06-10T16:43:25.052728242Z" -> (known after apply)
      ~ effective_labels = { # forces replacement
          + "foo" = "bar"
    }
```

I can't determine the root cause for this issue, this resource has a pre-update custom code that changes the updateMask to `*`, but even when I removed it, the issue still remains.

Hamzawy63 commented:
Gentle reminder @trodge

trodge commented:
> During my manual testing, I faced a strange behavior: If I make two consecutive update to the resource, the first update will be in-place (as expected) and the second one will force replace the resource. That was my resource definition:
> 
> ```
> resource "google_certificate_manager_trust_config" "default" {
>   name        = "sample-trust-configs"
>   description = "A sample trust config resource with allowlisted certificates"
>   location = "global"
> 
>   allowlisted_certificates  {
>     pem_certificate = file("cert.pem") 
>   }
>   labels = {
>     foo = "bar"
>   }
> }
> ```
> 
> The first update, I changed cert.pem to cert2.pem and the update was in-place and succssful The second update, I changed cert2.pem back to cert.pem. In this case, terraform decided to force replace the resource because it detected a change in effective_labels variable
> 
> ```
>   # google_certificate_manager_trust_config.default must be replaced
> -/+ resource "google_certificate_manager_trust_config" "default" {
>       ~ create_time      = "2024-06-10T16:43:25.052728242Z" -> (known after apply)
>       ~ effective_labels = { # forces replacement
>           + "foo" = "bar"
>     }
> ```
> 
> I can't determine the root cause for this issue, this resource has a pre-update custom code that changes the updateMask to `*`, but even when I removed it, the issue still remains.

It looks like in this resource the `effective_labels` field has `ForceNew`. I'm not sure why this would be since the field is computed. Removing it should fix this issue.

Hamzawy63 commented:
> It looks like in this resource the effective_labels field has ForceNew. I'm not sure why this would be since the field is computed. Removing it should fix this issue

Do you mean to make `labels` not immutable? as we don't have control over  `effective_labels` field (it does not belong to only TrustConfig resource).  See: https://github.com/GoogleCloudPlatform/magic-modules/pull/8950

I believe that making the `labels` not immutable will solve the issue, because `effective_labels` is immutable if `labels` is immutable. 

It's ok in this resource to make labels not immutable. There was an issue with updating labels before https://github.com/GoogleCloudPlatform/magic-modules/pull/8651#discussion_r1295239758, but it's no longer the case (I tested that manually). It's also ok to remove the `pre_update` custom code that was created earlier.

However, I want to point out two things:

- I'm surprised that `effective_labels` field changed during my experiment, I did not touch the labels during updating the resource and yet `effective_labels` changed for some reason. 
- I would guess that this is an issue that will face any resource that has an immutable `label` field (and hence an immutable `effective_label`).  Maybe it's rare to have a resource with immutable `labels` that's why this issue is not common.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 168 insertions(+), 5 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10906-old..auto-pr-10906) ( 3 files changed, 168 insertions(+), 5 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10906-old..auto-pr-10906) ( 1 file changed, 32 insertions(+))
Open in Cloud Shell: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 117 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: `20`
Passed tests: `17`
Skipped tests: `2`
Affected tests: `1`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>certificatemanager</li></ul></blockquote></details>


#### Action taken
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccCertificateManagerTrustConfig_update </blockquote></details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
#### Tests analytics
Total tests: `20`
Passed tests: `18`
Skipped tests: `2`
Affected tests: `0`

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>certificatemanager</li></ul></blockquote></details>

$\textcolor{green}{\textsf{All tests passed!}}$
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/ba7b4602-e007-4bdd-9bed-8626ce85acd3/build-log/replaying_test.log)

Hamzawy63 commented:
@trodge Friendly reminder

trodge commented on mmv1/products/certificatemanager/TrustConfig.yaml:
```diff
@@ -50,8 +50,11 @@ examples:
     vars:
       trust_config_name: 'trust-config'
-custom_code: !ruby/object:Provider::Terraform::CustomCode
-  pre_update: templates/terraform/pre_update/certificate_manager_trust_config.go.erb
```
Is this pre-update no longer required? It looks like it adds "*" as the update mask, is that no longer needed?

trodge commented on mmv1/products/certificatemanager/TrustConfig.yaml:
```diff
@@ -124,3 +126,15 @@ properties:
                   Each certificate provided in PEM format may occupy up to 5kB.
                 sensitive: true
+  - !ruby/object:Api::Type::Array
+    name: allowlistedCertificates
```
Could you also add a test that updates this field if it is updateable?

Hamzawy63 replied:
Done! added the field to the existing update test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 178 insertions(+), 7 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 178 insertions(+), 7 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-10906-old..auto-pr-10906) ( 1 file changed, 32 insertions(+))
Open in Cloud Shell: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-10906-old..auto-pr-10906) ( 4 files changed, 117 insertions(+))

modular-magician commented:
#### Tests analytics
Total tests: 20
Passed tests: 17
Skipped tests: 2
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>certificatemanager</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccCertificateManagerTrustConfig_update</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccCertificateManagerTrustConfig_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/7666ee71-2f84-4958-9cf8-4dcf109377a3/recording/TestAccCertificateManagerTrustConfig_update.log)]

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$

---




$\textcolor{green}{\textsf{All tests passed!}}$

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/7666ee71-2f84-4958-9cf8-4dcf109377a3/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-10906/artifacts/7666ee71-2f84-4958-9cf8-4dcf109377a3/recording) for each test
